File: BaekJoon/다이나믹 프로그래밍/1520. 내리막길.py
import sys
import heapq


def find(maps, dp, visited, n, m):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    queue = []

    heapq.heappush(queue, (maps[0][0], 0, 0))
    visited[0][0] = True
    dp[0][0] = 1

    while queue:
        count, x, y = heapq.heappop(queue)
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                if maps[x][y] > maps[nx][ny]:  # 다음 방문할 지점의 값이 현재의 위치보다 작은 경우
                    dp[nx][ny] += dp[x][y]
                    if not visited[nx][ny]:    # 방문하지 않았을 때만 큐에 넣어준다
                        visited[nx][ny] = True
                        heapq.heappush(queue, (-maps[nx][ny], nx, ny))

    return dp[n - 1][m - 1]


def main():
    n, m = map(int, input().split())
    maps = [[] for _ in range(n)]
    dp = [[0 for _ in range(m)] for _ in range(n)]
    visited = [[False for _ in range(m)] for _ in range(n)]

    for i in range(n):
        maps[i] = list(map(int, sys.stdin.readline().split()))

    print(find(maps, dp, visited, n, m))


if __name__ == "__main__":
    main()

File: BaekJoon/DP/1520_내리막 길.swift
let inputs = readLine()!.split(separator: " ").map { Int(String($0))! }
var maps = [[Int]]()

for _ in 0..<inputs[0] {
    maps.append(readLine()!.split(separator: " ").map { Int(String($0))! })
}

var dp = [[Int]](repeating: [Int](repeating: -1, count: inputs[1]), count: inputs[0])
let dx = [-1, 0, 1, 0]
let dy = [0, 1, 0, -1]

let ans = DFS(0, 0)
print(ans)

func DFS(_ x: Int, _ y: Int) -> Int {
    if x == inputs[0] - 1, y == inputs[1] - 1 { return 1 } // 기저
    if dp[x][y] != -1 { return dp[x][y] } // 이미 값을 갖고 있는 경우
    
    dp[x][y] = 0
    
    for i in 0..<4 {
        let nx = x + dx[i]
        let ny = y + dy[i]
        
        if nx >= 0, nx < inputs[0], ny >= 0, ny < inputs[1] {
            if maps[x][y] > maps[nx][ny] { // 다음 방문할 지점이 더 낮은 경우
                dp[x][y] += DFS(nx, ny) // 방문할 수 있는 경우의 수를 추가
            }
        }
    }
    
    return dp[x][y]
}